src/Searchbar.vue

```diff
@@ -15,38 +15,90 @@ export default {
+          if (!regex.test(info)) {
+            matched = false;
+          }
+        });
+        return matched;
+      }
+
       if (!this.data) {
         return undefined;
       }
       const matches = [];
-      const regex = new RegExp(this.value, 'i');
+      const regexes = this.value.split(' ').map(keyword => new RegExp(keyword, 'i'));
       this.data.forEach((entry) => {
         const { headings, src, title } = entry;
         const keywords = entry.keywords || '';
         let hasMatchingHeading = false;
+
         Object.entries(headings).forEach(([id, text]) => {
-          if (regex.test(text)) {
-            hasMatchingHeading = true;
-            matches.push({
-              heading: { id, text },
-              keywords,
-              src,
-              title,
-            });
-          }
-        });
-        if (!hasMatchingHeading) {
-          if (regex.test(title) || regex.test(keywords)) {
+            const info = [title].concat(keywords).concat(text).join(' ');
+            if (allMatch(info, regexes)) {
+              hasMatchingHeading = true;
+              matches.push({
+                heading: { id, text },
+                keywords,
+                src,
+                title,
+              });
+            }
+          });
+          if (!hasMatchingHeading) {
             matches.push(entry);
           }
-        }
       });
       return matches;
     },
   },
   filters: {
     highlight(value, phrase) {
-      return value.replace(new RegExp(`(${phrase})`, 'gi'), '<mark>$1</mark>');
+      function mergeIntervals(intervals) {
+        const stack = [];
+        let top = null;
+        if (intervals.length <= 1) {
+          return intervals;
+        }
+        const sortedIntervals = intervals.sort((a, b) => a.start - b.start);
+
+        stack.push(sortedIntervals[0]);
+        for (let i = 1; i < sortedIntervals.length; i += 1) {
+          top = stack[stack.length - 1];
+
+          if (top.end < sortedIntervals[i].start) {
+            stack.push(sortedIntervals[i]);
+          } else if (top.end < sortedIntervals[i].end) {
+            top.end = sortedIntervals[i].end;
+          }
+        }
+
+        return stack;
+      }
+
+      const keywords = phrase.split(' ').filter(keyword => keyword !== '');
+      let matchPos = [];
+      let match;
+      keywords.forEach((keyword) => {
+        const regex = new RegExp(`(${keyword})`, 'gi');
+        match = regex.exec(value);
+        while (match != null) {
+          matchPos.push({ start: match.index, end: match.index + keyword.length });
+          match = regex.exec(value);
+        }
+      });
+
+      matchPos = mergeIntervals(matchPos);
+
+      let result = value;
+      for (let i = matchPos.length - 1; i >= 0; i -= 1) {
+        result = `${result.slice(0, matchPos[i].start)}<mark>
+                  ${result.slice(matchPos[i].start, matchPos[i].end)}</mark>${result.slice(matchPos[i].end)}`;
```

Comment:
Doesn't `replace` work?

Reply:
I do agree that the algorithm seems quite complicated at first glance. Perhaps we can use a library to assist us in highlighting such as [mark.js](https://markjs.io/)?